Mathlib/Algebra/MonoidAlgebra/Degree.lean

```diff
@@ -222,35 +223,301 @@ variable [Semiring R]
 
 section SupDegree
 
-variable [AddZeroClass A] [SemilatticeSup B] [AddZeroClass B] [OrderBot B]
-  (D : A → B)
+variable [AddZeroClass A] [Add B] (D : A → B)
+
+section SemilatticeSup
+
+variable [SemilatticeSup B] [OrderBot B]
 
 /-- Let `R` be a semiring, let `A, B` be two `AddZeroClass`es, let `B` be an `OrderBot`,
 and let `D : A → B` be a "degree" function.
 For an element `f : R[A]`, the `R[A]`, the element `supDegree f : B` is the
 supremum of all the elements in the support of `f`, or `⊥` if `f` is zero.
 Often, the Type `B` is `WithBot A`,
 If, further, `A` has a linear order, then this notion coincides with the usual one,
-using the maximum of the exponents. -/
-@[reducible]
-def supDegree (f : R[A]) : B :=
-  f.support.sup D
+using the maximum of the exponents.
+
+If `A := σ →₀ ℕ` then `R[A] = MvPolynomial σ R`, and if we equip `σ` with a linear order then
+the induced linear order on `Lex A` equips `MvPolynomial` ring with a
+[monomial order](https://en.wikipedia.org/wiki/Monomial_order) (i.e. a linear order on `A`, the
+type of (monic) monomials in `R[A]`, that respects addition). We make use of this monomial order
+by taking `D := toLex`, and different monomial orders could be accessed via different type
+synonyms once they are added. -/
+@[reducible] def supDegree (f : R[A]) : B := f.support.sup D
 
-theorem supDegree_add_le (f g : R[A]) :
+/-- If `D` is an injection into a linear order `B`, the leading coefficient of `f : R[A]` is the
+  nonzero coefficient of highest degree according to `D`, or 0 if `f = 0`. In general, it is defined
+  to be the coefficient at an inverse image of `supDegree f` (if such exists). -/
+noncomputable def leadingCoeff (f : R[A]) : R := f (D.invFun <| f.supDegree D)
```

Comment:
Do you have a use-case in mind where you don't know the inverse of `D`? If not, then I think taking an `Equiv` would make more sense here

Reply:
@adomani may have some use cases in mind where `D := WithBot.some` (injective but not Equiv), which would produce something like `Polynomial.degree` rather than `natDegree`. This is necessary for Laurent polynomials, where `A = ℤ` doesn't have a `bot` so we'd have to adjoin one to use `Finset.sup`. When `D` is not injective then the current definition is not well-behaved and shouldn't be used IMO.

Reply:
It's also not that well behaved for `d = id` on `A = Prod Nat Nat`, right? The leading coefficient of `X + Y + 1` is `0`, since the `sup` gives the coefficient of `X*Y`?

Reply:
Yeah, I should have said `leadingCoeff` is also not intended to be used when `B` isn't a linear order.

Reply:
I think if that's the case we should probably just write that assumption for the lemmas about `leadingCoeff`

Reply:
Currently, only four lemmas `leadingCoeff_single, leadingCoeff_zero, Monic.ne_zero, monic_one` about `leadingCoeff` and `Monic` don't assume `LinearOrder`. Do you mean that you want to replace `SemilatticeSup B` in the definitions of `leadingCoeff` and `Monic` by `LinearOrder B`?

Reply:
Yes, I think we should change those definitions to assume `LinearOrder B`

Reply:
I'll change `leadingCoeff` and `Monic` to use `LinearOrder` but I think I'll keep D a plain function. `Equiv` wouldn't be convenient for the [weighted order](https://github.com/leanprover-community/mathlib4/pull/7173#discussion_r1329081072) I mentioned and for Laurent polynomials. I'll add two lemmas saying that you can use any left inverse to `D` in place of `D.invFun` if `D` is injective, and that when `D` is an `Equiv` you can use `D.symm`. (We could consider bundling injectivity of D into `Embedding` but that doesn't seem convenient either.)